pkg/controller/template/render.go

```diff
@@ -391,9 +396,38 @@ func skipMissing(key string) (interface{}, error) {
 	return fmt.Sprintf("{{.%s}}", key), nil
 }
 
+// isCloudProviderExternal is a wrapper for the library-go IsCloudProviderExternal function, it is needed
+// to address an issue related to the vSphere in-tree storage driver and the CSI migration feature gate.
+// On vSphere, this function will check the storage operator configuration to determine if the in-tree
+// driver is in use, if it is this function will return false. Adding this function in the MCO allows
+// us to have a configuration whereby the kubelet can use the in-tree cloud controller logic to configure
+// itself, while allowing the kube-apiserver and kube-controller-manager to use the external cloud controllers.
+// For more information about the root issue, please see the following links:
+// https://github.com/kubernetes/kubernetes/pull/116342
+// https://issues.redhat.com/browse/STOR-1265
+func isCloudProviderExternal(platformStatus *configv1.PlatformStatus, featureGate *configv1.FeatureGate, storageConfig *operatorv1.Storage) (bool, error) {
+	if platformStatus == nil {
+		return false, fmt.Errorf("platformStatus is required")
+	}
+	switch platformStatus.Type {
+	case configv1.VSpherePlatformType:
+		// Platforms that are external based on feature gate presence
+		external, err := cloudprovider.IsCloudProviderExternal(platformStatus, featureGate)
+		if err != nil {
+			return external, nil
+		}
+		isexternal := external && ((storageConfig == nil) || (storageConfig != nil && storageConfig.Spec.VSphereStorageDriver == operatorv1.CSIWithMigrationDriver))
```

Comment:
i'm not sure if this is the proper logic for cases where `storageConfig == nil`, will we ever have a condition where that storage api is non-existent?

cc @jsafrane @dobsonj

Reply:
thanks @jsafrane !